libsignal-protocol/src/crypto.rs

```diff
@@ -56,49 +299,185 @@ impl CryptoProvider {
 
 struct State(Box<dyn Crypto>);
 
+unsafe extern "C" fn random_func(
+    data: *mut u8,
+    len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!data.is_null());
+    assert!(!user_data.is_null());
+
+    let user_data = &*(user_data as *const State);
+    let buffer = slice::from_raw_parts_mut(data, len);
+    user_data.0.fill_random(buffer).into_code()
+}
+
 unsafe extern "C" fn hmac_sha256_cleanup_func(
     _hmac_context: *mut c_void,
-    _user_data: *mut c_void,
+    user_data: *mut c_void,
 ) {
-    unimplemented!();
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    user_data.0.hmac_sha256_cleanup();
 }
 
 unsafe extern "C" fn hmac_sha256_final_func(
     _hmac_context: *mut c_void,
-    _output: *mut *mut signal_buffer,
-    _user_data: *mut c_void,
+    output: *mut *mut signal_buffer,
+    user_data: *mut c_void,
 ) -> i32 {
-    unimplemented!()
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    match user_data.0.hmac_sha256_final() {
+        Ok(buf) => {
+            let buffer = Buffer::from(buf);
+            output.write(buffer.into_raw());
```

Comment:
Ah, it took me a couple seconds of staring to realise this is just `*output = buffer.into_raw()`. 

I usually associate `write()` with the `std::io::Write` interface, so I thought it was trying to copy `buffer`'s contents into `output` and then leak `buffer` :laughing:

If you do an `assert!(!output.is_null())` at the start, that should be enough. This *is* an `unsafe` function being called by C, so we need to trust that C has given us something valid.

Reply:
Isn't `output` is an empty buffer ? Isn't that means it is `null` ? I'm a bit confused, since i didn't check out the c implementation of the `signal_buffer`

Reply:
It's a little confusing, but `output` is actually a pointer to a pointer to a `signal_buffer` :laughing: 

The [double-pointer thing](https://stackoverflow.com/a/897400) is usually used when you need to return a pointer to some memory on the heap, but not using the return value (e.g. you also need to return an error code). In this case, the C library is asking *us* to allocate a new buffer, and it's passing in a pointer to their `*mut signal_buffer` so we can make it point to the newly allocated buffer.

In C, this might look like:

```rust
int ERR_NO_MEM = -1;
int SUCCESS = 0;

int allocate_hello_world(char **buffer) {
  // allocate our string
  char *s = malloc(42);

  // check for errors
  if (s == NULL) {
    return ERR_NO_MEM;
  }
  
  // copy "Hello, World" into the allocated buffer
  strcpy(s, "Hello, World!");

  // Everything was successful, make the pointer passed into the function
  // point to our allocated string
  *buffer = s;
  return SUCCESS;
}

int main() {
  char *message;
  // Pass message by reference we want the function to change what it points to
  int ret = allocate_hello_world(&message);
  
  if (ret != SUCCESS) {
    // bail
  }

  printf("%s\n", message);
}
```

Reply:
yeah, that's interesting, thank you for the clarification :heart: